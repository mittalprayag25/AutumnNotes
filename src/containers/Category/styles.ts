import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    height: '100%',
  },
  header: {
    marginBottom: 16,
    fontSize: 24,
  },
  topContainer: {
    minHeight: 128,
  },
  editNote: {
    height: '20%',
    borderColor: '#abc',
    padding: 16,
    borderWidth: 2,
    marginHorizontal: 16,
  },
  listItem: {
    padding: 5,
    margin: 10,
    alignItems: 'flex-start',
  },
  navContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  categoryLabel: {
    marginLeft: 8,
  },
  title: {
    fontSize: 24,
  },
  listContainer: {
    marginHorizontal: 16,
  },
  category: {
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    backgroundColor: '#eeeee4',
    margin: 8,
    borderRadius: 10,
    paddingHorizontal: 10,
  },
});

export default styles;
