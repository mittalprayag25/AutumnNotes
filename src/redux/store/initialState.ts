const initialState = {
  notes: {
    allNotes: [],
    singleNote: null,
  },
  categories: {
    categories: [],
  },
};

export default initialState;
