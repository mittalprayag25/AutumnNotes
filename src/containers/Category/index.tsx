import {
  Text,
  Image,
  View,
  TouchableOpacity,
  TextInput,
  FlatList,
} from 'react-native';
import React, {useState} from 'react';
import HeaderNav from '../../components/HeaderNav';
import {useDispatch, useSelector} from 'react-redux';
import images from '../../assets/index';
import actions from '../../redux/actions';
import styles from './styles';

const Category = ({navigation}: any) => {
  const [category, setCategory] = useState(null);
  const dispatch = useDispatch();
  const {categories}: any = useSelector((s: RootState) => s);

  const saveCategory = () => {
    if (category) {
      dispatch(actions.categories.createcategory({category}));
      setCategory(null);
    }
  };

  const createCategoryObject = rawCategory => {
    const obj: any = {};
    obj.label = rawCategory;
    obj.value = rawCategory;
    return obj;
  };

  const removeCategory = (data: any, index: number) => {
    dispatch(actions.categories.deletecategory({category: data, index}));
  };

  const renderItem = ({item, index}) => (
    <View style={styles.category}>
      <Text style={styles.listItem}>{`${item.value}`}</Text>
      <Text
        testID={`create-category-remove-${item.value}`}
        onPress={text => {
          removeCategory(text, index);
        }}>
        x
      </Text>
    </View>
  );

  const addCategory = item => {
    setCategory(createCategoryObject(item));
  };

  const renderLeftnav = () => (
    <>
      <TouchableOpacity
        onPress={() => navigation.goBack()}
        testID="create-category-back">
        <Image source={images.notes.back} />
      </TouchableOpacity>
    </>
  );

  const renderRightNav = () => (
    <>
      <TouchableOpacity
        onPress={() => saveCategory()}
        testID="create-category-save">
        <Image source={images.notes.save} />
      </TouchableOpacity>
    </>
  );

  return (
    <>
      <View style={styles.container}>
        <HeaderNav
          leftNav={renderLeftnav()}
          rightNav={renderRightNav()}
          title="Create category"
        />
        <View style={styles.container}>
          <View style={styles.editNote}>
            <TextInput
              value={category ? category : ''}
              onChangeText={addCategory}
              testID="create-category-input-text"
              multiline={true}
              autoFocus
              selectionColor="#000"
            />
          </View>
          <FlatList
            style={styles.listContainer}
            data={categories.categories}
            renderItem={renderItem}
            numColumns={3}
          />
        </View>
      </View>
    </>
  );
};

export default Category;
