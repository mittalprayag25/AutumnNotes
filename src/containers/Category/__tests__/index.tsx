import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import * as reactRedux from 'react-redux';
import {Provider} from 'react-redux';

import {
  createMockStore,
  mockNavigationWithParams,
} from '../../../__mocks__/index';
import Category from '../index';
import initialState from '../../../redux/store/initialState';

const mockedDispatch = jest.fn();

describe('Create Category', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      categories: {
        categories: [
          {label: 'cat1', value: 'cat1'},
          {label: 'cat2', value: 'cat2'},
        ],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Category {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should save when text input value is there', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Category {...props} />
      </Provider>,
    );
    const textInput = getByTestId('create-category-input-text');
    const save = getByTestId(`create-category-save`);
    act(() => {
      fireEvent(save, 'press');
    });
    expect(mockedDispatch).not.toHaveBeenCalled();

    fireEvent.changeText(textInput, '98756789');

    act(() => {
      fireEvent(save, 'press');
    });
    expect(mockedDispatch).toHaveBeenCalled();
  });
  it('should go back', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Category {...props} />
      </Provider>,
    );
    const goBack = getByTestId('create-category-back');
    act(() => {
      fireEvent(goBack, 'press');
    });
    expect(navigation.goBack).toHaveBeenCalledTimes(1);
  });
  it('should remove category', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Category {...props} />
      </Provider>,
    );
    const text = getByTestId('create-category-remove-cat2');
    act(() => {
      fireEvent(text, 'press');
    });
    expect(mockedDispatch).toHaveBeenCalled();
  });
});
