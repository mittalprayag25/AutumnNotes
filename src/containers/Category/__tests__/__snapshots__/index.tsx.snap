// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Category should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="create-category-back"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/left.png",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Create category
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="create-category-save"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/save.png",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "#abc",
          "borderWidth": 2,
          "height": "20%",
          "marginHorizontal": 16,
          "padding": 16,
        }
      }
    >
      <TextInput
        autoFocus={true}
        multiline={true}
        onChangeText={[Function]}
        selectionColor="#000"
        testID="create-category-input-text"
        value=""
      />
    </View>
    <RCTScrollView
      data={
        Array [
          Object {
            "label": "cat1",
            "value": "cat1",
          },
          Object {
            "label": "cat2",
            "value": "cat2",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginHorizontal": 16,
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#eeeee4",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignItems": "flex-start",
                    "margin": 10,
                    "padding": 5,
                  }
                }
              >
                cat1
              </Text>
              <Text
                onPress={[Function]}
                testID="create-category-remove-cat1"
              >
                x
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#eeeee4",
                  "borderRadius": 10,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignItems": "flex-start",
                    "margin": 10,
                    "padding": 5,
                  }
                }
              >
                cat2
              </Text>
              <Text
                onPress={[Function]}
                testID="create-category-remove-cat2"
              >
                x
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
